### DISCUSSION TITLE
Vercel - NextJS rewrites not working as expected in production

### DISCUSSION THREAD

**Discussion by radikris** [vercel/vercel#8393] [Help]
https://stackoverflow.com/questions/73345094/vercel-nextjs-rewrites-not-working-as-expected-in-production

I have a website built by Next.js and we deployed it on Vercel.

We localize our website in 5 different languages, and as I found the solution with rewrites for localizing the urls as well.

We have a custom page: customer-stories, and would like to localize these:
 

      customer_stories: {
    hu: '/esettanulmanyok',
    en: '/customer-stories',
    de: '/kundengeschichte',
    sk: '/pribehy-zakaznikov',
    cs: '/pribehy-zakazniku',
  },

So in the next config file:
Inside rewrites:

     {
        source: '/hu/esettanulmanyok/:id*/',

        destination: '/hu/customer-stories/:id*/',
        locale: false,
      },
      {
        source: '/de/kundengeschichte/:id*/',
        destination: '/de/customer-stories/:id*/',
        locale: false,
      },

      {
        source: '/cs/pribehy-zakazniku/:id*/',
        destination: '/cs/customer-stories/:id*/',
        locale: false,
      },
      {
        source: '/sk/pribehy-zakaznikov/:id*/',
        destination: '/sk/customer-stories/:id*/',
        locale: false,
      },

Which is working perfectly on localhost:
http://localhost:3000/hu/esettanulmanyok/
[![enter image description here][1]][1]

And a dynamic id page:
http://localhost:3000/hu/esettanulmanyok/newtestwithreference/
[![enter image description here][2]][2]

They work as expected, but on the released site it is different:
The same url is 404: https://barion-builderio.vercel.app/hu/esettanulmanyok/
[![enter image description here][3]][3]

The same dynamic url is working: https://barion-builderio.vercel.app/hu/esettanulmanyok/newtestwithreference/
[![enter image description here][4]][4]

What is very interested, because the main page is using pagination with ***getInitialProps***:
When I go the original page without rewrites: https://barion-builderio.vercel.app/hu/customer-stories/
[![enter image description here][5]][5]

And after that, I paginate on the website:
[![enter image description here][6]][6]
And click on the next page:
[![enter image description here][7]][7]

The url is changing to the correct one, but if I try to refresh on the page, again 404:
[![enter image description here][8]][8]

So the dynamic [id] site is working well with rewrites, but the list page with getinitialprops not working for the first time, only with the original url, and after that if the user uses the pagination the rewrites will working. On the page we use shallow routing for not loading the site again.

I don't know why is it working properly on localhost and not on the server:
Here is the full (important part) code of my customer-stories.js:

        imports...
    
    const articlesPerPage = 4
    
    const CustomerStories = ({
      stories,
      texts,
      tophero,
      bottomhero,
      enumlocalization,
      page,
      navbar,
      footer,
      subfooter,
    }) => {
      const { locale, pathname } = useRouter()
    
      const [currentPage, setCurrentPage] = useState(page)
    
    
      const [filteredData, setFilteredData] = useState(stories)
    
      const router = useRouter()
    
    
      const handlePageReset = () => {
        console.log('reset futott')
        setCurrentPage(0)
        handlePaginationClick(0)
      }
    
      const {
        country,
        industry,
        solution,
        handleIndustry,
        handleCountry,
        handleSolution,
      } = useCustomerStoriesContext()
    
    	//FILTERING ON THE SITE
      useDidMountEffect(async () => {
        const queryArray = {}
    
        if (solution && solution.key != null) {
          queryArray = {
            ...queryArray,
            'data.solution.$elemMatch.solution.type': solution.key.type,
          }
        }
        if (industry && industry.key != null) {
          queryArray = {
            ...queryArray,
            'data.industry.type.$eq': industry.key.type,
          }
        }
        if (country && country.key != null) {
          queryArray = {
            ...queryArray,
            'data.country.type.$eq': country.key.type,
          }
        }
    
        const result = await builder.getAll('barion-userstory-data', {
          options: {
            noTargeting: true,
            query: queryArray,
            page: 0,
          },
    
          omit: 'data.blocks',
        })
        handlePageReset()
        setFilteredData((prev) => result)
      }, [country, industry, solution])
    
      const handlePaginationClick = (page) => {
        //setCurrentPage(page)
    
        router.push(
          {
            pathname: LOCALIZED_URLS['customer_stories'][locale],
            query: page != 0 ? { page: page != 0 ? page : '' } : {},
          },
          undefined,
          { shallow: false }
        )
      }
    
      return (
        <>
          <Head>
            <meta name="robots" content="noindex" />
          </Head>
         <div>
    		PAGE LIST CONTENT CUSTOMER-STORIES
          </div>
        </>
      )
    }
    
    CustomerStories.getInitialProps = async ({
      req,
      res,
      asPath,
      query,
      locale,
    }) => {
      if (locale == 'default') {
        return {
          props: {
            plugins: null,
            texts: null,
            enumlocalization: null,
            integrationready: null,
          },
        }
      }
    
      
    	... FETCH DATA
    
      const page = Number(query.page) || 0
    
      let stories = await builder.getAll('barion-userstory-data', {
        options: {
          noTargeting: true,
          // sort: {  //TODO ADD SORTING ALPHABETICALLY
          //   'data.title': 1,
          // },
          offset: page * articlesPerPage,
        },
        limit: articlesPerPage * 3,
        omit: 'data.blocks',
      })
    
    ...
    
      return {
        stories,
        tophero,
        bottomhero,
        texts: texts?.data.texts[locale],
        enumlocalization: enumlocalization.data,
        page,
        navbar,
        footer,
        subfooter,
      }
    }
    
    export default CustomerStories

  [1]: https://i.stack.imgur.com/eSuIi.png
  [2]: https://i.stack.imgur.com/HjrAR.png
  [3]: https://i.stack.imgur.com/RiRf3.png
  [4]: https://i.stack.imgur.com/rbV3h.png
  [5]: https://i.stack.imgur.com/lgVIZ.png
  [6]: https://i.stack.imgur.com/wO5ZM.png
  [7]: https://i.stack.imgur.com/T9WQo.png
  [8]: https://i.stack.imgur.com/lW4FM.png

**Comment by amyegan** [comment]
Thanks for reaching out @radikris. Am I correct in assuming that you already tried the [Rewrites with basePath support](https://nextjs.org/docs/api-reference/next.config.js/rewrites#rewrites-with-basepath-support) option and it isn't working for you?
